src/main/java/org/broadinstitute/hellbender/utils/codecs/TargetCodec.java

```diff
@@ -34,9 +38,63 @@ public Target decode(final String line) {
         return sourceTargetTableReader.readRecord(line);
     }
 
+    /**
+     * TargetCodec uses a com.opencsv.CSVReader, which uses a buffered reader. Since just initializing the
+     * CSVReader causes it to buffer a large portion of the underlying input stream, it can't be used directly
+     * on an input stream that is being indexed.
+     *
+     * To avoid this, we override makeIndexableSourceFromStream and instantiate and return a marker class that is a
+     * subclass of AsciiLineReaderIterator, and which is recognized by the readActualHeader method (where the actual
+     * TargetTableReader/CSVReader state is established). The marker class signals the codec to manually extract
+     * everything up to the end of the header from the stream, and instantiate the CSVReader on a separate stream
+     * containing only the (header) portion of the input. This allows the CSV reader to correctly establish it's
+     * internal state based on the header, without consuming the rest of the stream.
+     *
+     * The rest of the indexing process doesn't delegate the remaining input stream contents to the codec (the
+     * indexer parses the stream manually, and passes the resulting features to the codec's decode method), so
+     * the features are then consumed and indexed with the correct stream offsets.
+     *
+     * @param inputStream stream to be indexed
+     * @return An AsciiLineReaderIterator implementation.
+     */
+    @Override
+    public LocationAware makeIndexableSourceFromStream(final InputStream inputStream) {
+        if (inputStream instanceof BlockCompressedInputStream) {
```

Comment:
The fact that we have to duplicate this logic here makes me lean towards Magic's suggestion of creating a factory method in AsciiLineReader that encapsulates this in one place.